git/versioning-faq.md

```diff
@@ -0,0 +1,140 @@
+
+--------------------------------------------------------------------------------
+
+### Committing
+
+There are many different philosophies and practices for version control. This
+document provides some guidance on what practices we implement and why.
+
+#### What does a good commit message look like?
+
+> 1. Separate subject from body with a blank line
+> 1. Limit the subject line to 50 characters
+> 1. Capitalize the subject line
+> 1. Do not end the subject line with a period
+> 1. Use the imperative mood in the subject line
+> 1. Wrap the body at 72 characters
+> 1. Use the body to explain what and why vs. how
+
+Source: https://chris.beams.io/posts/git-commit/
+
+__Have meaningful commits__
+
+Other people including your future self will be reading your history. Make them
+useful.
+
+An example of what not to do:
+- "Save work"
+- "Save more work"
+- "Finished feature"
+
+Further Reading: [Commit Message Guide](./commit_message_guide.md)
+
+#### How big should a commit be?
+
+A commit should be a logical chunk of work. The meaning of this can vary from
+team to team, so make sure you are all aligned.
+
+> A properly formed Git commit subject line should always be able to complete
+the following sentence:
+> - If applied, this commit will __your subject line here__
+
+This can be used as a helpful guiding principle of how big a commit should be.
+
+For example a helpful commit could look like
+- "Implement the login screen"
+- "Extract a class"
+- "Update dependencies"
+- "Fix typo on landing page"
+- "Apply look and feel to the landing page"
+
+In the same regard, if you cannot succinctly describe that sentence, then your
+commit might be too small or too large.
+
+__Break down large commits__
+- "Add the login page, onboarding flow, settings page, and networking layer"
+
+This commit is large! You could consider instead breaking this down into four
+smaller commits
+- "Implement the networking layer"
+- "Implement the login page"
+- "Implement the onboarding flow"
+- "Implement the settings page"
+
+__Combine small commits__
+- "Add function to modify text"
+- "Add test for function"
+- "Connect user properties and labels"
+
+Instead this could be:
+- "Hook up the user profile page"
+
+#### When should I commit?
+
+Similar to the above, you should commit when you have created a logical chunk of
+work. If you are actively working, you will likely want to commit at least
+several times a day.
+
+If you practice rewriting history, commit as frequently as you desire as long as
+history looks good in the end.
+
+--------------------------------------------------------------------------------
+
+### Workflow
+
+For further information on this see [this sherpa page.](https://github.com/
+IntrepidPursuits/sherpa/blob/master/git/git_workflow_quick_reference.md)
+
+#### When should I branch?
+Generally speaking, when you need to work in parallel with another ongoing set
+of work.
```

Comment:
I feel that this is a little *too* lenient. The correct answer is *always*. 

But in all seriousness, this implies that one person, working solo, can work on `master` or `develop` or whatever. But even *then* there should be a distinct branch, for PR review purposes if nothing else.

Reply:
Not sure I still have write access here.

@bobgilmore how about instead just saying:
```
#### When should I branch?
Almost always.
... <list as is>
```

Reply:
Looks good to me, for what that's worth :-) Not certain that I have write access, either. Maybe I'll try to modify the branch tomorrow and see what happens ;-)